src/parameterFiles.ts

```diff
@@ -0,0 +1,507 @@
+      case yes.title:
+        await setMappedParamFileForTemplate(templateUri, closestMatch.uri);
+        break;
+      case no.title:
+        // We won't ask again
+        await neverAskAgain(templateUri, actionContext);
+
+        // Let them know how to do it manually
+        // Don't wait for an answer
+        window.showInformationMessage(howToMessage);
+        break;
+      case another.title:
+        await commands.executeCommand("azurerm-vscode-tools.selectParameterFile", templateUri);
+        break;
+      default:
+        assert("considerQueryingForParameterFile: Unexpected response");
+        break;
+    }
+  });
+}
+
+function canAsk(templateUri: Uri, actionContext: IActionContext): boolean {
+  const checkForMatchingParamFilesSetting: boolean = !!workspace.getConfiguration(configPrefix).get<boolean>(configKeys.checkForMatchingParamFiles);
+  actionContext.telemetry.properties.checkForMatchingParamFiles = String(checkForMatchingParamFilesSetting);
+  if (!checkForMatchingParamFilesSetting) {
+    return false;
+  }
+
+  // tslint:disable-next-line: strict-boolean-expressions
+  const neverAskFiles: string[] = ext.context.globalState.get<string[]>(globalStateKeys.dontAskAboutParamFiles) || [];
+  const key = normalizePath(templateUri);
+  if (neverAskFiles.includes(key)) {
+    actionContext.telemetry.properties.isInDontAskList = 'true';
+    return false;
+  }
+
+  let ignoreThisSession = _filesToIgnoreThisSession.has(normalizePath(templateUri));
+  if (ignoreThisSession) {
+    actionContext.telemetry.properties.ignoreThisSession = 'true';
+    return false;
+  }
+
+  return true;
+}
+
+function dontAskAgainThisSession(templateUri: Uri, actionContext: IActionContext): void {
+  _filesToIgnoreThisSession.add(normalizePath(templateUri));
+}
+
+async function neverAskAgain(templateUri: Uri, actionContext: IActionContext): Promise<void> {
+  // tslint:disable-next-line: strict-boolean-expressions
+  const neverAskFiles: string[] = ext.context.globalState.get<string[]>(globalStateKeys.dontAskAboutParamFiles) || [];
+  const key: string = normalizePath(templateUri);
+  neverAskFiles.push(key);
+  await ext.context.globalState.update(globalStateKeys.dontAskAboutParamFiles, neverAskFiles);
+}
+
+/**
+ * Given a template file, find the parameter file, if any, that the user currently has associated with it
+ */
+export function findMappedParamFileForTemplate(templateFileUri: Uri): Uri | undefined {
+  const paramFiles: { [key: string]: string } | undefined =
+    workspace.getConfiguration(configPrefix).get<{ [key: string]: string }>(configKeys.parameterFiles)
+    // tslint:disable-next-line: strict-boolean-expressions
+    || {};
+  if (typeof paramFiles === 'object') {
+    const normalizedTemplatePath = normalizePath(templateFileUri.fsPath);
+    let paramFile: Uri | undefined;
+    for (let fileNameKey of Object.getOwnPropertyNames(paramFiles)) {
+      const normalizedFileName: string | undefined = normalizePath(fileNameKey);
```

Comment:
>normalizePath [](start = 53, length = 13)

Don't we save them normalized? Do we need to do this again? Also, do we need the loop - can we just lookup property value via property name [propertyName]?

Reply:
Ah, nevermind, looks like we don't save them normalized. Makes sense I guess since that's what the user will be seeing. Even on Windows they would want case to match to what's in the filesystem.

---
In reply to: [389204136](https://github.com/microsoft/vscode-azurearmtools/pull/514#discussion_r389204136) [](ancestors = 389204136)

Reply:
No, this is from user settings and thus modifiable by the user.  Can't do a simple lookup because the case might be different.